File: src/main/java/com/itszuvalex/femtocraft/power/CryogenRegistry.java
package com.itszuvalex.femtocraft.power;

import net.minecraft.world.World;

import java.util.ArrayList;

/**
 * Created by Chris on 8/19/2014.
 */
public class CryogenRegistry {
    private static ArrayList<ICryogenHandler> passiveHandlers = new ArrayList<ICryogenHandler>();
    private static ArrayList<ICryogenHandler> activeHandlers = new ArrayList<ICryogenHandler>();

    public static void init() {
        registerPassiveHandler(new CryogenPassiveHandler());
        registerActiveHandler(new CryogenActiveHandler());
    }

    public static boolean registerPassiveHandler(ICryogenHandler handler) {
        return passiveHandlers.add(handler);
    }

    public static boolean registerActiveHandler(ICryogenHandler handler) {
        return activeHandlers.add(handler);
    }

    public static boolean unregisterPassiveHandler(ICryogenHandler handler) {
        return passiveHandlers.remove(handler);
    }

    public static boolean unregisterActiveHandler(ICryogenHandler handler) {
        return activeHandlers.remove(handler);
    }

    public static float getPassivePower(World world, int x, int y, int z) {
        return getPower(world, x, y, z, passiveHandlers);
    }

    public static float getActivePower(World world, int x, int y, int z) {
        return getPower(world, x, y, z, activeHandlers);
    }

    private static float getPower(World world, int x, int y, int z, ArrayList<ICryogenHandler> handlers) {
        for (ICryogenHandler handler : handlers) {
            if (handler.canHandleBlock(world, x, y, z)) {
                float amount = handler.powerForBlock(world, x, y, z);
                handler.usedBlockForPower(world, x, y, z);
                return amount;
            }
        }
        return 0;
    }
}

File: src/main/java/com/itszuvalex/femtocraft/power/CryogenRegistry.scala
package com.itszuvalex.femtocraft.power

import com.itszuvalex.femtocraft.api.power.ICryogenHandler
import net.minecraft.world.World

import scala.collection.mutable.ArrayBuffer

/**
 * Created by Chris on 8/19/2014.
 */
object CryogenRegistry {
  private val passiveHandlers = new ArrayBuffer[ICryogenHandler]
  private val activeHandlers  = new ArrayBuffer[ICryogenHandler]

  def init() {
    registerPassiveHandler(new CryogenPassiveHandler)
    registerActiveHandler(new CryogenActiveHandler)
  }

  def registerPassiveHandler(handler: ICryogenHandler) = passiveHandlers += handler

  def registerActiveHandler(handler: ICryogenHandler) = activeHandlers += handler

  def unregisterPassiveHandler(handler: ICryogenHandler) = passiveHandlers -= handler

  def unregisterActiveHandler(handler: ICryogenHandler) = activeHandlers -= handler

  def getPassivePower(world: World, x: Int, y: Int, z: Int) = getPower(world, x, y, z, passiveHandlers)

  def getActivePower(world: World, x: Int, y: Int, z: Int) = getPower(world, x, y, z, activeHandlers)

  private def getPower(world: World, x: Int, y: Int, z: Int, handlers: ArrayBuffer[ICryogenHandler]): Float = {
    handlers.filter(_.canHandleBlock(world, x, y, z)).foreach { handler =>
      val amount = handler.powerForBlock(world, x, y, z)
      handler.usedBlockForPower(world, x, y, z)
      return amount
                                                              }
    0
  }
}

